materialize-bigquery/binding.go

```diff
@@ -1,225 +1,369 @@
-	store struct {
-		paramsConverter sqlDriver.ParametersConverter
-		extDataConfig   *bigquery.ExternalDataConfig
-		mergeFile       *ExternalDataConnectionFile
-		sql             string
-		tempTableName   string // The name the external table we be referenced by
-		hasRootDocument bool   // Whether the root document is included in this binding
-	}
+type writer struct {
+	dataCommitted bool
+	dataWritten   int64
+	edc           bigquery.ExternalDataConfig
+	object        *storage.ObjectHandle
+	writer        *storage.Writer
+	buffer        *bufio.Writer
+	encoder       *json.Encoder
 }
 
-// bindingDocument is used by the load operation to fetch binding flow_document values
-type bindingDocument struct {
-	Binding  int
-	Document json.RawMessage
+type sqlQueries struct {
+	LoadSQL  string
+	WriteSQL string
 }
 
-// Load implements the bigquery ValueLoader interface for mapping data from the database to this struct
-// It requires 2 values, the first must be the binding and the second must be the flow_document
-func (bd *bindingDocument) Load(v []bigquery.Value, s bigquery.Schema) error {
+var externalTableNameTemplate = "external_table_binding_%s"
 
-	if len(v) != 2 {
-		return fmt.Errorf("invalid value count: %d", len(v))
+// Returns a new compiled Binding ready to be used. The Binding is generated from a materialization spec and merges the config
+// info available to generate data structure that only requires to be compiled once when the materialization starts up. When
+// this struct is called, all the information to generate the proper SQL queries is already available.
+// The Binding object also includes a mutable part that is encapsulated inside the `writer` struct.
+func NewBinding(ctx context.Context, cfg *config, bucket *storage.BucketHandle, bindingSpec *pf.MaterializationSpec_Binding) (*Binding, error) {
+	schema, err := schemaForBinding(bindingSpec)
+	if err != nil {
+		return nil, err
 	}
-	if binding, ok := v[0].(int64); !ok {
-		return fmt.Errorf("value[0] wrong type %T expecting int", v[0])
-	} else {
-		bd.Binding = int(binding)
+
+	var br bindingResource
+
+	if err := pf.UnmarshalStrict(bindingSpec.ResourceSpecJson, &br); err != nil {
+		return nil, fmt.Errorf("parsing resource config: %w", err)
 	}
-	if document, ok := v[1].(string); !ok {
-		return fmt.Errorf("value[1] wrong type %T expecting string", v[0])
-	} else {
-		bd.Document = json.RawMessage(document)
+
+	externalTableAlias := fmt.Sprintf(externalTableNameTemplate, br.Table)
+
+	b := &Binding{
+		Queries:            generateSQLQueries(externalTableAlias, br, bindingSpec),
+		bucket:             bucket,
+		spec:               bindingSpec,
+		schema:             schema,
+		region:             cfg.Region,
+		dataset:            cfg.Dataset,
+		externalTableAlias: externalTableAlias,
 	}
 
-	return nil
+	name, err := randomString()
+	if err != nil {
+		return nil, err
+	}
 
+	b.Reset(ctx, fmt.Sprintf("%s/%s", cfg.BucketPath, name))
```

Comment:
Just to clarify: are we basically initialising a new writer on an object with a random name here? This is the writer used to write the file corresponding to this binding yeah?

It may be a nitpick but can we somehow express the purpose of this function with a different name? I see that it's called whenever we want a new writer. So perhaps just `binding.NewStorageWriter` or something along those lines

Reply:
I think that makes sense, I'll make the change.